### PR TITLE
support cp ,fix qwen3.5 gdn sp

### PR DESCRIPTION
**Pull request by meichangsu1**
# PR type
- [x] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

This PR adds context parallel and Qwen3.5 Gated DeltaNet sequence parallel support to the transformers stack, and refactors sequence parallel into a package-based implementation.

Main changes:
- Refactor `sequence_parallel.py` into `sequence_parallel/` and add shared utilities.
- Add derived ring / zigzag ring attention support for CP + SP.
- Add Qwen3.5 linear attention SP support in `linear_attention_sp.py`;Ring attention is not supported for this path yet.
- Update transformers model / processor paths to work with the new SP+CP flow.
- Adjust loss metric aggregation for Ulysses replicated loss behavior.
- Update cookbook examples for `sp_fsdp_dense`.
- Add test coverage for:
  - Qwen3.5 linear attention SP alignment
  - sequence parallel + context parallel behavior
- Remove outdated `tests/moe/test_expert_parallel_qwen3_fsdp_sp.py`.

### CODE REVIEW

File: src/twinkle/metric/loss.py [hunk 1]
```diff
@@ -25,13 +25,19 @@ def accumulate(self, inputs: Union[InputFeature, List[InputFeature]], outputs: M
...
                 inputs = [inputs]
             for input in inputs:
                 # `Transformers` models may use reduction=sum, to average grads before step
                 labels = input['labels']
                 self.num_tokens += (labels >= 0).sum().item()
+            # Sequence-parallel gathered loss is replicated on each ulysses rank, while
+            # local labels still count only the shard-local tokens. Normalize the loss
+            # contribution here so metric-side averaging matches the non-SP path.
+            if ulysses_size > 1:
+                loss = loss / float(ulysses_size)
```
Comment (tastelikefeet) [inline src/twinkle/metric/loss.py]: 为什么会放到这里呢，或者说，model进行backward的loss是否需要除以ulysses-size

File: src/twinkle/model/transformers/strategy/sequence_parallel/linear_attention_sp.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+from transformers.utils.import_utils import is_flash_linear_attention_available
+from typing import Any, Optional, Tuple
+
+from twinkle.model.transformers.strategy.sequence_parallel.utils import head_to_seq_shard, seq_to_head_shard
+from twinkle.patch import Patch
+
+if is_flash_linear_attention_available():
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/strategy/sequence_parallel/linear_attention_sp.py:10]: 这部分能使用到swift中吗？swift也缺linear-attention的sp

File: src/twinkle/model/transformers/transformers.py [hunk 3]
```diff
@@ -42,6 +42,18 @@
 from twinkle.utils.grad_clip import normalize_and_clip_grad_norm
 
 
+def _get_raw_dp_fsdp_world_size(device_mesh: Optional[DeviceMesh]) -> int:
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这个和device_mesh的dp_world_size似乎是一样的？能否复用

File: src/twinkle/model/transformers/transformers.py [hunk 4]
```diff
@@ -467,7 +470,11 @@ def calculate_loss(self, **kwargs):
         inputs = optimizer_config.train_status.inputs
         outputs = optimizer_config.train_status.outputs
         assert inputs is not None and outputs is not None, 'Cannot calculate loss of empty inputs and outputs'
-        result = loss_instance(inputs, outputs, **kwargs)
+        loss_inputs = inputs
+        loss_outputs = outputs
+        if self.sp_strategy is not None:
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这部分能否使用inputprocessor？既然切分是inputprocessor做，那gather是否应该也放在里面

File: src/twinkle/model/transformers/strategy/sequence_parallel/linear_attention_sp.py (hunk 2 above)
Comment (meichangsu1) [reply src/twinkle/model/transformers/strategy/sequence_parallel/linear_attention_sp.py:10]: 可以的

File: src/twinkle/model/transformers/transformers.py (hunk 3 above)
Comment (meichangsu1) [reply src/twinkle/model/transformers/transformers.py]: 不一样，这里算的是 dp_world_size * fsdp_world_size，device_mesh的dp_world_size是：    @property
    def dp_world_size(self) -> int:
        return self._get_world_size_for_dim('dp')

File: src/twinkle/model/transformers/transformers.py (hunk 4 above)
Comment (meichangsu1) [reply src/twinkle/model/transformers/transformers.py]: 应该不太合适吧，这里已经是到了loss 计算阶段了，inputprocessor的职责应该是做输入的处理的吧

File: src/twinkle/metric/loss.py (hunk 1 above)
Comment (meichangsu1) [reply src/twinkle/metric/loss.py]: loss_instance 的reduction 为sum时，这里loss 是 在每个 ulysses rank 上都复制了一份的 全序列 loss，但这里统计的 num_tokens 还是 每个 rank 本地 shard 的 token 数。两边口径不一致，所以要除一次 ulysses_size，这里除一下只是只为修 metric 打印口径；至于反向传播时loss是没有除以ulysses size的，在GatherLoss.apply中只保留了本地梯度

Comment (meichangsu1) [reply src/twinkle/metric/loss.py]: 还有我感觉这里当开启sp时这里这么判断会有点问题，当fsdp=2时，raw_dp_world_size=2,而data_world_size=1,此时就跳过gather了
df3eb2f42ce7411a076b820ffd3372f0

另外这里应该也要改成raw_dp_fsdp_world_size,因为后面gather 的维度是process_group,不是data_world_size
16b985153d0f9675f6fcbad778eba2cb



image

File: src/twinkle/model/transformers/transformers.py (hunk 3 above)
Comment (tastelikefeet) [reply src/twinkle/model/transformers/transformers.py]: Image
你看看这个实现满足需求吗

Comment (tastelikefeet) [reply src/twinkle/model/transformers/transformers.py]: 这个和data_rank是专门用来判断数据组的，而且即使真的增加dpworldsize，实现也应该放在devicemesh里面复用，而非放在transformers.py里面，模型和数据组判断耦合会带来维护问题

File: src/twinkle/model/transformers/transformers.py (hunk 4 above)
Comment (tastelikefeet) [reply src/twinkle/model/transformers/transformers.py]: inputprocessor的名字起的可能不太好，这个组件就是为了做任务相关的数据处理的，放在模型代码里面，如果再增加一个子类，实现不是要重写一遍

**Review by tastelikefeet — APPROVED** [review]
